src/account-info/AccountDetailsScreen.js

```diff
@@ -23,28 +23,50 @@ const styles = StyleSheet.create({
     fontStyle: 'italic',
     fontSize: 18,
   },
+
+  userNotFoundScreen: {
+    flex: 1,
+    justifyContent: 'center',
+    alignItems: 'center',
+  },
 });
 
 type SelectorProps = $ReadOnly<{|
   isActive: boolean,
-  user: UserOrBot,
+  user: UserOrBot | void,
 |}>;
 
 type Props = $ReadOnly<{|
-  navigation: NavigationScreenProp<{ params: {| email: string |} }>,
+  navigation: NavigationScreenProp<{ params: {| userId: number | void |} }>,
 
   dispatch: Dispatch,
   ...SelectorProps,
 |}>;
 
 class AccountDetailsScreen extends PureComponent<Props> {
-  handleChatPress = () => {
-    const { user, dispatch } = this.props;
-    dispatch(doNarrow(privateNarrow(user.email)));
+  componentDidMount() {
+    const { user } = this.props;
+    if (user === undefined) {
+      logging.error('User object passed is undefined.');
+    }
+  }
+
+  handleChatPress = (email: string) => {
+    const { dispatch } = this.props;
+    dispatch(doNarrow(privateNarrow(email)));
   };
 
   render() {
     const { user, isActive } = this.props;
+
+    if (user === undefined) {
+      return (
+        <Screen title="Profile not found" style={styles.userNotFoundScreen}>
+          <Label text="The profile for this user could not be found." />
+        </Screen>
+      );
+    }
+
```

Comment:
So, one thing that's contributing quite a bit of the complexity of the code in this PR is the way it's handling the possibility of being asked to show a user that doesn't actually exist. Specifically, there's about four different spots in this component's code that deal with `props.user` potentially being undefined instead of a real User object, in addition to the places here and elsewhere dealing with potentially even the `userId` navigation parameter being undefined.

When lots of different bits of code all try to respond to the same error condition, it makes it hard to understand what the code is doing either in that error case or in the normal case, and hard to edit or refactor the code. So let's try to make sure this case gets taken care of once, as early as possible, and then all the code downstream of that can just deal with a real User object.

This principle is basically a particular form of the ["crunchy shell, soft center"](https://github.com/zulip/zulip-mobile/blob/master/docs/architecture/crunchy-shell.md) principle.

In particular this means `props.user` should go on being just `UserOrBot`, not `UserOrBot | void`, so that all the code inside the component just deals with a real, valid user object.

Specifically, what I'd like to do right now in this code is:
* If `userId` (or `email`, if applicable in the earlier parts of the branch) doesn't actually find us a user object in our data structures, just throw an error.
  * See `getStreamForId` for an example of how to do this conveniently. We can add a similar selector `getUserForId`.
* Similarly, over in `InfoNavButtonPrivate`, if the given user doesn't actually exist in our data then just throw an error.
  * A selector `getUserForEmail`, on the same lines as above, may be helpful for this.

Throwing an error is appropriate because I believe it's the case that if we actually have an invalid user ID or email here, that's a bug in the app.

If it later turns out that there is a legitimate case where that can happen -- or if such a bug does exist, and we don't simply debug and fix it, and it's happening with some frequency -- then we can add a bit more complexity to give a nice user-facing error like this "user not found" screen. If we do that, the key principle above will still be an important one: only one spot of code should have to handle the error case, letting everything else deal only with real, valid data.
* One likely way to do that would be to have a little wrapper component, which shows either the "not found" screen or the actual account-details screen. The wrapper's `user` prop would be potentially undefined, but the inner component's `user` prop would not be.

Reply:
Thanks for the detailed explanations! I too felt the changes were getting messy.